File: 백준/Gold/16437. 양 구출 작전/양 구출 작전.cc
#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);

using namespace std;

struct node{
    bool is_sheep = true;
    int val = 0;
};

node nodes[123'457];
vector<int> edges[123'457];

long long calc(int num, int parent){
    long long result = 0ll;

    for(int child : edges[num]){
        if(child == parent) continue;

        result += calc(child, num);
    }

    if(nodes[num].is_sheep) result += nodes[num].val;
    else result = max(0ll, result - nodes[num].val);

    return result;
}

int main(){
    FASTIO

    int n;
    cin >> n;

    for(int i = 2; i <= n; i++){
        char t;
        int a, p;
        cin >> t >> a >> p;

        nodes[i] = {(t == 'S'), a};
        edges[i].push_back(p);
        edges[p].push_back(i);
    }

    cout << calc(1, -1);

    return 0;
}

File: 백준/Gold/16437. 양 구출 작전/양 구출 작전.py
import sys
sys.setrecursionlimit(10**6)

n = int(sys.stdin.readline())

tree = [[] for _ in range(n + 1)]
wolf_sheep = [[0, 0] for _ in range(n + 1)]

for i in range(n - 1):
    t, a, p = sys.stdin.readline().split()
    tree[int(p)].append(i + 2)
    wolf_sheep[i + 2] = [t, int(a)]

def dfs(node, parent):
    result = 0
    for child in tree[node]:
        if child == parent:
            continue
        result += dfs(child, node)
    
    if wolf_sheep[node][0] == 'W':
        result = max(0, result - wolf_sheep[node][1])
    else:
        result += wolf_sheep[node][1]
    
    return result

print(dfs(1, 0))